File: src/main/java/ludwig/interpreter/Interpreter.java
package ludwig.interpreter;

import ludwig.model.*;
import org.pcollections.HashPMap;
import org.pcollections.HashTreePMap;

public class Interpreter {
    public static Object eval(Node node, HashPMap<NamedNode, Object> locals) {
        return node.accept(new Evaluator(locals));
    }

    public static Object call(FunctionNode functionNode, Object... args) {
        ReferenceNode head = new ReferenceNode(functionNode);

        for (Object arg : args) {
            head.add(LiteralNode.Companion.ofValue(arg));
        }

        return eval(head, HashTreePMap.empty());
    }

}

File: src/main/kotlin/ludwig/interpreter/Interpreter.kt
package ludwig.interpreter

import ludwig.model.*
import org.pcollections.HashPMap
import org.pcollections.HashTreePMap

object Interpreter {
    fun eval(node: Node, locals: HashPMap<NamedNode, Any>): Any? {
        return node.accept(Evaluator(locals))
    }

    fun call(functionNode: FunctionNode, vararg args: Any?): Any? {
        val head = ReferenceNode(functionNode)

        for (arg in args) {
            head.children.add(LiteralNode.ofValue(arg!!))
        }

        return eval(head, HashTreePMap.empty())
    }

}
